identity-iota/src/credential/errors.rs

```diff
@@ -0,0 +1,127 @@
+  /// correct.
+  #[error("the credential's structure is not semantically correct")]
+  CredentialStructure(#[source] identity_credential::Error),
+  /// Indicates that the structure of the [Presentation](identity_credential::presentation::Presentation) is not
+  /// semantically correct.
+  #[error("the presentation's structure is not semantically correct")]
+  PresentationStructure(#[source] identity_credential::Error),
+  /// Indicates that the relationship between the presentation holder and one of the credential subjects is not valid.
+  #[error("expected holder = subject of the credential at position {credential_position}")]
+  #[non_exhaustive]
+  InvalidHolderSubjectRelationship { credential_position: usize },
+  /// Indicates that the presentation does not have a holder.
+  #[error("the presentation has an empty holder property")]
+  MissingPresentationHolder,
+}
+
+// Todo: Consider implementing Error for OneOrMany<E: std::error::Error> to avoid wrapping it in
+// AccumulatedCredentialValidationError
+#[derive(Debug)]
+/// An error caused by a failure to validate a Credential.  
+pub struct AccumulatedCredentialValidationError {
+  pub validation_errors: OneOrMany<ValidationError>,
+}
+
+impl Display for AccumulatedCredentialValidationError {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    // intersperse might become available in the standard library soon: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.intersperse
+    let detailed_information: String = itertools::intersperse(
+      self.validation_errors.iter().map(|err| err.to_string()),
+      ", ".to_string(),
+    )
+    .collect();
+    write!(f, "The following errors occurred: {}", detailed_information)
+  }
+}
+
+impl std::error::Error for AccumulatedCredentialValidationError {}
+
+#[derive(Debug)]
+/// An error caused by a failure to validate a Presentation.
+pub struct AccumulatedPresentationValidationError {
+  pub credential_errors: BTreeMap<usize, AccumulatedCredentialValidationError>,
+  pub presentation_validation_errors: Vec<ValidationError>,
+}
+
+impl Display for AccumulatedPresentationValidationError {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    let credential_error_formatter = |(position, reason): (&usize, &AccumulatedCredentialValidationError)| -> String {
+      format!(
+        "could not validate credential at position {}. The following errors occurred {}",
+        position,
+        reason.to_string().as_str()
+      )
+    };
+
+    let error_string_iter = self
+      .presentation_validation_errors
+      .iter()
+      .map(|error| error.to_string())
+      .chain(self.credential_errors.iter().map(credential_error_formatter));
+    let detailed_information: String = itertools::intersperse(error_string_iter, ", ".to_string()).collect();
+    write!(f, "the following errors occurred: {}", detailed_information)
```

Comment:
Not really sure how this looks when logged, since it will also be wrapped in e.g. `IotaError`.

Reply:
here is an example:

The following code: 
```
 let error = validator
    .validate(&credential, &options, &other_issuer_resolved_doc, fail_fast)
    .unwrap_err(); 
    let report = anyhow::Error::from(error);
    println!("anyhow report without causes:");
    println!("{}", report);
    println!("");
    println!("anyhow report with causes:");
    println!("{:#}", report);
```
produces the following output. 

```
anyhow report without causes:
credential validation failed

anyhow report with causes:
credential validation failed: The following errors occurred: issuance date is in the future or later than required, the expiration date is in the past or earlier than required, the credential's structure is not semantically correct, issuer did not match any of the supplied DID Documents
```

Reply:
I changed this display implementation a bit. Here is an example of what it looks like now:

The following code 
```
    let report = anyhow::Error::from(error);
    println!("anyhow report without causes:");
    println!("{}", report);
    println!("");
    println!("anyhow report with causes:");
    println!("{:#}", report);
```
produces the following output:
```
anyhow report without causes:
presentation validation failed

anyhow report with causes:
presentation validation failed: [expected holder = subject of the credential at position 0; holder does not match provided DID Document; the credential at position 0 failed validation for the following reasons: [issuance date is in the future or later than required; the expiration date is in the past or earlier than required]; the credential at position 1 failed validation for the following reasons: [issuance date is in the future or later than required; the expiration date is in the past or earlier than required]]
```

Reply:
I think it's okay. Some of the errors could maybe be more brief.

E.g. `the credential at position 0 failed validation for the following reasons:`
could just be `credential 0 errors:` or something similar.